### PR TITLE
Add polygons obstacles

### PR DESCRIPTION
**Pull request by agustinesco**
The main goal of this pr is implement the logic for when a player is colliding with a polygon, the circle one is already implemented so we will part from there

This pr will implement the [SAT (Separated Axis Algorith)](https://dyn4j.org/2010/01/sat/) to get the amount of movement and direction of the collision but only when the previous implemented detection get that we collided with a polygon since it's cheaper

## Disclaimer: This solution will only work for Convex polygons, it means that we can't have something like an U or V, a solution for this would be making two polygons.

I'll list some shapes that i've tested myself, to use them simply paste one of the shapes into the `obstacles` list present in the `config.json` file and either run the game in unity or go to [This](http://localhost:3000/board/1/uma/player_name) link in your browser while the backend is running:

<details>
<summary>Wrong Four point star with a concave poligon</summary>

``` json
 {
        "position": {
          "x": 0.0,
          "y": 0.0
        },
        "vertices": [
          {
            "x": 0.0,
            "y": 1000.0
          },
          {
            "x": 250.0,
            "y": 250.0
          },
          {
            "x": 1000.0,
            "y": 0.0
          },
          {
            "x": 250.0,
            "y": -250.0
          },
          {
            "x": 0.0,
            "y": -1000.0
          },
          {
            "x": -250.0,
            "y": -250.0
          },
          {
            "x": -1000.0,
            "y": 0.0
          },
          {
            "x": -250.0,
            "y": 250.0
          }
        ],
        "radius": 0.0,
        "shape": "polygon"
      }
```
</details>

<details>
<summary>Correct Four point star made from several convex poligons</summary>

``` json
{
    "position": {
        "x": 0.0,
        "y": 0.0
    },
    "vertices": [
        {
            "x": 0.0,
            "y": 250.0
        },
        {
            "x": 1000.0,
            "y": 0.0
        },
        {
            "x": 0.0,
            "y": -250.0
        }
    ],
    "radius": 0.0,
    "shape": "polygon"
},
{
    "position": {
        "x": 0.0,
        "y": 0.0
    },
    "vertices": [
        {
            "x": 0.0,
            "y": 250.0
        },
        {
            "x": -1000.0,
            "y": 0.0
        },
        {
            "x": 0.0,
            "y": -250.0
        }
    ],
    "radius": 0.0,
    "shape": "polygon"
},
{
    "position": {
        "x": 0.0,
        "y": 0.0
    },
    "vertices": [
        {
            "x": -250.0,
            "y": 0.0
        },
        {
            "x": 0.0,
            "y": -1000.0
        },
        {
            "x": 250.0,
            "y": 0.0
        }
    ],
    "radius": 0.0,
    "shape": "polygon"
},
{
    "position": {
        "x": 0.0,
        "y": 0.0
    },
    "vertices": [
        {
            "x": -250.0,
            "y": 0.0
        },
        {
            "x": 0.0,
            "y": 1000.0
        },
        {
            "x": 250.0,
            "y": 0.0
        }
    ],
    "radius": 0.0,
    "shape": "polygon"
},
```
</details>

<details>
<summary>Diamond</summary>

``` json
       {
        "position": {
          "x": 0.0,
          "y": 0.0
        },
        "vertices": [
          {
            "x": 0.0,
            "y": 1000.0
          },
          {
            "x": 1000.0,
            "y": 0.0
          },
          {
            "x": 0.0,
            "y": -1000.0
          },
          {
            "x": -1000.0,
            "y": 0.0
          }
        ],
        "radius": 0.0,
        "shape": "polygon"
      }
```
</details>

<details>
<summary>Emerald</summary>

``` json
     {
        "position": {
          "x": 0.0,
          "y": 0.0
        },
        "vertices": [
          {
            "x": 0.0,
            "y": 3000.0
          },
          {
            "x": 2000.0,
            "y": 1000.0
          },
          {
            "x": 2000.0,
            "y": -1000.0
          },
          {
            "x": 0.0,
            "y": -3000.0
          },
          {
            "x": -2000.0,
            "y": -1000.0
          },
          {
            "x": -2000.0,
            "y": 1000.0
          }
        ],
        "radius": 0.0,
        "shape": "polygon"
      }
```
</details>

<details>
<summary>Concave shape</summary>

``` json
    {
        "position": {
          "x": -250.0,
          "y": 0.0
        },
        "vertices": [
          {
            "x": -1000.0,
            "y": 500.0
          },
          {
            "x": 0.0,
            "y": 250.0
          },
          {
            "x": 0.0,
            "y": -250.0
          },
          {
            "x": -1000.0,
            "y": -250.0
          }
        ],
        "radius": 0.0,
        "shape": "polygon"
      },
      {
        "position": {
          "x": 250.0,
          "y": 0.0
        },
        "vertices": [
          {
            "x": 1000.0,
            "y": 500.0
          },
          {
            "x": 0.0,
            "y": 250.0
          },
          {
            "x": 0.0,
            "y": -250.0
          },
          {
            "x": 1000.0,
            "y": -250.0
          }
        ],
        "radius": 0.0,
        "shape": "polygon"
      }
```
</details>

<details>
<summary>Cross</summary>

``` json
    {
    "position": {
        "x": 0.0,
        "y": 0.0
    },
    "vertices": [
        {
            "x": 0.0,
            "y": 250.0
        },
        {
            "x": 1000.0,
            "y": 250.0
        },
        {
            "x": 1000.0,
            "y": -250.0
        },
        {
            "x": 0.0,
            "y": -250.0
        }
    ],
    "radius": 0.0,
    "shape": "polygon"
},
{
    "position": {
        "x": 0.0,
        "y": 0.0
    },
    "vertices": [
        {
            "x": 0.0,
            "y": 250.0
        },
        {
            "x": -1000.0,
            "y": 250.0
        },
        {
            "x": -1000.0,
            "y": -250.0
        },
        {
            "x": 0.0,
            "y": -250.0
        }
    ],
    "radius": 0.0,
    "shape": "polygon"
},
{
    "position": {
        "x": 0.0,
        "y": 0.0
    },
    "vertices": [
        {
            "x": -250.0,
            "y": 0.0
        },
        {
            "x": -250.0,
            "y": -1000.0
        },
        {
            "x": 250.0,
            "y": -1000.0
        },
        {
            "x": 250.0,
            "y": 0.0
        }
    ],
    "radius": 0.0,
    "shape": "polygon"
},
{
    "position": {
        "x": 0.0,
        "y": 0.0
    },
    "vertices": [
        {
            "x": -250.0,
            "y": 0.0
        },
        {
            "x": -250.0,
            "y": 1000.0
        },
        {
            "x": 250.0,
            "y": 1000.0
        },
        {
            "x": 250.0,
            "y": 0.0
        }
    ],
    "radius": 0.0,
    "shape": "polygon"
},
```
</details>

### CODE REVIEW

File: apps/arena/priv/config.json [hunk 1]
```diff
@@ -49,21 +49,116 @@
...
       {
         "position": {
           "x": -1630.0,
           "y": 4588.0
         },
-        "radius": 317.0
+        "radius": 317.0,
+        "shape": "circle"
+      },
+      {
```
Comment (manucamejo) [inline apps/arena/priv/config.json:71]: Is there a PR on the front? I think we should remove (or maybe comment) the obstacles until the client has a representation for them.

**Review by manucamejo — APPROVED** [review]

**Comment by manucamejo** [comment]
@agustinesco please add some documentation in our docs.

File: apps/arena/priv/config.json (hunk 1 above)
Comment (agustinesco) [reply apps/arena/priv/config.json]: whoops, yes this was something that we added to test on a playtest some days ago. i deleted this obstacles

**Comment by agustinesco** [comment]
Documentation added! @manucamejo

**Review by ncontinanza — APPROVED** [review]
LGTM!
I guess this is for a future PR, but take into account that currently a player or an item may spawn inside an obstacle

image

File: apps/arena/docs/collision-detection.md [hunk 2]
```diff
@@ -1,5 +1,6 @@
 # Collision detection
 
+## Jeffrey thompson implementation
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:3]:
```suggestion
## Jeffrey Thompson's implementation
```

File: apps/arena/docs/collision-detection.md [hunk 3]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
 
 It is necessary to clarify that there are other types of collisions that we do not need to implement at the moment, for example Polygon/Polygon.
 
+## SAT (Simple Axis Theorem) implamentation
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:58]:
```suggestion
## SAT (Simple Axis Theorem) implementation
```

File: apps/arena/docs/collision-detection.md [hunk 4]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+A collision detection library implementation based on the [SAT theorem](https://dyn4j.org/2010/01/sat/).
+
+We'll keep working with entities with the same shapes as the Jeffrey thompson
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:61]:
```suggestion
We'll keep working with entities with the same shapes as the ones used by Jeffrey Thompson.
```

File: apps/arena/docs/collision-detection.md [hunk 5]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+
+To detect a collision between two entities we'll iterate over an axis for each pair of vertices from both entities that collided,
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:63]:
```suggestion
To detect a collision between two entities, we'll iterate over an axis for each pair of vertices from both entities that collided.
```

File: apps/arena/docs/collision-detection.md [hunk 6]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+for circle shaper we'll check the axis that's formed between the circle center and the closes polygon vertex, and then cast the
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:64]:
```suggestion
For circular shapes, we'll check the axis that's formed between the circle center and the closest polygon vertex, and then cast the
```

File: apps/arena/docs/collision-detection.md [hunk 7]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+maximun and minimun vertex to the normal of each pair of vertices. As soon as we find an axis where the cast of the shapes are not overlapping we can safely say that the entities are not colliding
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:65]:
```suggestion
maximum and minimum vertex to the normal of each pair of vertices. As soon as we find an axis where the cast of the shapes are not overlapping, we can safely say that the entities are not colliding.
```

File: apps/arena/docs/collision-detection.md [hunk 8]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+
+#### Tiangle example for axis where the cast in the axis are overlapping
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:67]:
```suggestion
#### Triangle example for axis where the cast in the axis are overlapping
```

File: apps/arena/docs/collision-detection.md [hunk 9]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+![Polygons axis overlapping](./images/sat-overlapping.jpg "Polygons axis overlapping")
+#### Tiangle example for axis where the cast in the axis are not overlapping
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:69]:
```suggestion
#### Triangle example for axis where the cast in the axis are not overlapping
```

File: apps/arena/docs/collision-detection.md [hunk 10]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+![Polygons axis not overlapping](./images/sat-no-overlapping.jpg "Polygons axis not overlapping")
+
+The collisions we curretly support are:
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:72]:
```suggestion
The collisions we currently support are:
```

File: apps/arena/docs/collision-detection.md [hunk 11]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+1. Circle with polygon:
+![Circle/Polygon not colliding](./images/circle-polygon-not-colliding.jpg "Circle/Polygon not colliding")
+
+![Circle/Polygon colliding](./images/circle-polygon-colliding.jpg "Circle/Polygon colliding")
+
+
+
+## Collision resolution
+
+The Jeffrey thompson implementation for collision detection is ideal since it's the cheapest way to
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:83]:
```suggestion
The Jeffrey Thompson's implementation for collision detection is ideal since it's the cheapest way to 
```

File: apps/arena/docs/collision-detection.md [hunk 12]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+![Circle/Polygon not colliding](./images/circle-polygon-not-colliding.jpg "Circle/Polygon not colliding")
+
+![Circle/Polygon colliding](./images/circle-polygon-colliding.jpg "Circle/Polygon colliding")
+
+
+
+## Collision resolution
+
+The Jeffrey thompson implementation for collision detection is ideal since it's the cheapest way to 
+detect that two entities collided, but it's incomplete we can only detect that a collision occurred
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:84]:
```suggestion
detect that two entities collided, but it's incomplete since we can only detect that a collision occurred 
```

File: apps/arena/docs/collision-detection.md [hunk 13]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+
+![Circle/Polygon colliding](./images/circle-polygon-colliding.jpg "Circle/Polygon colliding")
+
+
+
+## Collision resolution
+
+The Jeffrey thompson implementation for collision detection is ideal since it's the cheapest way to 
+detect that two entities collided, but it's incomplete we can only detect that a collision occurred 
+nothing else, so we decided to take the following approach:
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:85]:
```suggestion
and nothing else, so we decided to take the following approach:
```

File: apps/arena/docs/collision-detection.md [hunk 14]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+
+1. We use the Jeffrey thompson implementation to detect that a collision actually happened
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:87]:
```suggestion
1. We use the Jeffrey Thompson's implementation to detect that a collision actually happened.
```

File: apps/arena/docs/collision-detection.md [hunk 15]
```diff
@@ -54,3 +55,35 @@ If you only need to detect when a circle collides with the "walls" of the polygo
...
+2. Using SAT we can get the normal that will resolve that collision and the amount of movement we should do 
+so the entities are no longer colliding, so when a collision occur we will use sat to push the entities in the direction that, with the minimum movement, will resolve that collision
```
Comment (ncontinanza) [inline apps/arena/docs/collision-detection.md:89]:
```suggestion
so the entities are no longer colliding, so when a collision occurs we will use SAT to push the entities in the direction that, with the minimum movement, will resolve that collision.
```

**Review by ncontinanza — CHANGES_REQUESTED** [review]

**Review by ncontinanza — APPROVED** [review]

**Review by Nico-Sanchez — CHANGES_REQUESTED** [review]
Collisions feel REALLY NICE. Congratulations!

I left a lot of commends regarding the codebase of this feature. I think it must be polished in the following aspects:
1. Code reusability. Several functions are implemented for Position type but they aren't used in some cases where they could.
2. Code readability. I know SAT takes time to understand, but the code is really hard to follow. Some variable and function naming could be more friendly.
3. Documentation. There are many typos or sentences that could be more explicit or clear. Try to think of someone without the context to think if it's good enough.

After polishing this we're ready to test them live 🤩

File: apps/arena/lib/arena/entities.ex [hunk 16]
```diff
@@ -261,4 +268,10 @@ defmodule Arena.Entities do
   def maybe_add_custom_info(_entity) do
     nil
   end
+
+  defp get_shape("polygon"), do: :polygon
+  defp get_shape("circle"), do: :circle
+  defp get_shape("line"), do: :line
+  defp get_shape("point"), do: :point
+  defp get_shape(_), do: nil
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/lib/arena/entities.ex:276]: It's a default case so this doesn't explode 💥

File: apps/arena/lib/arena/game_updater.ex [hunk 17]
```diff
@@ -609,7 +611,13 @@ defmodule Arena.GameUpdater do
         Map.put(
           obstacles_acc,
           last_id,
-          Entities.new_circular_obstacle(last_id, obstacle.position, obstacle.radius)
+          Entities.new_obstacle(
+            last_id,
+            obstacle.position,
+            obstacle.radius,
+            obstacle.shape,
+            obstacle[:vertices]
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/lib/arena/game_updater.ex:619]: this won't do , if you use the dot to access to a variable in the obstacle map it'll crash if it doesn't have said variable, you can check it yourself in the iex:

```
iex(1)> map = %{radius: 2}
%{radius: 2}
iex(2)> map.vertices
** (KeyError) key :vertices not found in: %{radius: 2}

iex(2)> map[:vertices]
nil
iex(3)>
```

File: apps/arena/native/physics/src/collision_detection.rs [hunk 18]
```diff
@@ -1,5 +1,5 @@
 use crate::map::{Entity, Position};
-
+pub mod sat;
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/collision_detection.rs:2]: This is used to define the sad module in rust

File: apps/arena/native/physics/src/map.rs [hunk 19]
```diff
@@ -55,6 +55,33 @@ pub enum Category {
     Item,
 }
 
+impl Position {
+    pub fn normalize(&mut self) {
+        let len = (self.x.powi(2) + self.y.powi(2)).sqrt();
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/map.rs]: yes! i just renamed it

File: apps/arena/native/physics/src/map.rs [hunk 20]
```diff
@@ -55,6 +55,33 @@ pub enum Category {
...
+            x: a.x - b.x,
+            y: a.y - b.y,
+        }
+    }
+
+    pub fn distance_to_position(&self, other_position: &Position) -> f32 {
+        let x = self.x - other_position.x;
+        let y = self.y - other_position.y;
+        (x.powi(2) + y.powi(2)).sqrt()
+    }
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/map.rs:82]: hmmmm maybe there's a misunderstanding about the normalizing and the distance , on normalize we are getting the length of the vertex, and here we're getting distance between positions

File: apps/arena/native/physics/src/map.rs [hunk 21]
```diff
@@ -172,21 +199,41 @@ impl Entity {
...
-            y: entity.position.y
-                + normalized_direction.y * entity.radius
-                + normalized_direction.y * self.radius,
+    pub fn move_to_edge_position_outside(&mut self, collisions: Vec<&Entity>) {
+        for entity in collisions {
+            match entity.shape {
+                Shape::Circle => {
+                    let x = self.position.x - entity.position.x;
+                    let y = self.position.y - entity.position.y;
+                    let length = (x.powf(2.) + y.powf(2.)).sqrt();
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/map.rs:208]: hmm i dont know about this code, this was something that was already implemented, i'll change it to use the new normalize from position

File: apps/arena/native/physics/src/map.rs [hunk 22]
```diff
@@ -172,21 +199,41 @@ impl Entity {
...
+                    let normalized_direction = Position {
+                        x: x / length,
+                        y: y / length,
+                    };
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/map.rs:212]: yeah, i used the normalize/1 and sub/2 from position instead of this code!

File: apps/arena/priv/config.json [hunk 23]
```diff
@@ -3,11 +3,11 @@
     "radius": 4800.0,
     "initial_positions": [
       {
-        "x": 4000.0,
+        "x": 2000.0,
         "y": 0.0
       },
       {
-        "x": -4000.0,
+        "x": -2000.0,
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/priv/config.json:10]: whoops, restored! i think this was something left from the playtest changes

File: apps/arena/native/physics/src/collision_detection/sat.rs [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+    let mut axis: Position;
+
+    // Check normal and depth for polygon
+    for current in 0..polygon.vertices.len() {
+        let mut next = current + 1;
+        if next == polygon.vertices.len() {
+            next = 0
+        };
+        let va = polygon.vertices[current];
+        let vb = polygon.vertices[next];
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/collision_detection/sat.rs]: Changed!

File: apps/arena/native/physics/src/collision_detection/sat.rs [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+) -> (bool, Position, f32) {
+    // The normal will be the vector in wich the polygons should move to stop colliding
+    let mut normal = Position { x: 0.0, y: 0.0 };
+    // The depth is the amount of overlapping between both entities
+    let mut depth: f32 = f32::MAX;
+
+    let mut axis: Position;
+
+    // Check normal and depth for polygon
+    for current in 0..polygon.vertices.len() {
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/collision_detection/sat.rs]: done

File: apps/arena/native/physics/src/collision_detection/sat.rs [hunk 26]
```diff
@@ -0,0 +1,268 @@
...
+        // the axis will be the perpendicular line drawn from the current line of the polygon
+        axis = Position {
+            x: -current_line.y,
+            y: current_line.x,
+        };
+
+        // FIXME normalizing on this loop may be bad
+        axis.normalize();
+        let (min_a, max_a) = project_vertices(&polygon.vertices, axis);
+        let (min_b, max_b) = project_circle(circle, axis);
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/collision_detection/sat.rs]: sure thing! this was something out of desperation 😂

File: apps/arena/native/physics/src/collision_detection/sat.rs [hunk 27]
```diff
@@ -0,0 +1,268 @@
...
+            if depth_b > depth_a {
+                normal = Position {
+                    x: -axis.x,
+                    y: -axis.y,
+                };
+            } else {
+                normal = axis;
+            }
+            depth = axis_depth;
+        }
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/collision_detection/sat.rs]: changed

File: apps/arena/native/physics/src/collision_detection/sat.rs [hunk 28]
```diff
@@ -0,0 +1,268 @@
...
+        if depth_b > depth_a {
+            normal = Position {
+                x: -axis.x,
+                y: -axis.y,
+            };
+        } else {
+            normal = axis;
+        }
+        depth = axis_depth;
+    }
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/collision_detection/sat.rs]: i don't get what's hard to follow, can you explain it a little bit?

**Comment by agustinesco** [comment]
> Documentation. There are many typos or sentences that could be more explicit or clear. Try to think of someone without the context to think if it's good enough.

About this, the documentation is not meant to explain sat in a deep level, it's just enough to explain what the specific piece of code is doing. Since i've already linked a deep dive in what SAT is and the logic behind it you should first read that if you want to really understand what is happening

**Comment by Nico-Sanchez** [comment]
> > Documentation. There are many typos or sentences that could be more explicit or clear. Try to think of someone without the context to think if it's good enough.
> 
> About this, the documentation is not meant to explain sat in a deep level, it's just enough to explain what the specific piece of code is doing. Since i've already linked a deep dive in what SAT is and the logic behind it you should first read that if you want to really understand what is happening

Yeah, I totally agree. I meant regarding the code comments that you already wrote.

File: apps/arena/native/physics/src/map.rs [hunk 29]
```diff
@@ -163,20 +163,39 @@ impl Entity {
...
+                }
+            }
+            Shape::Polygon => {
+                let x = self.position.x - entity.position.x;
+                let y = self.position.y - entity.position.y;
+                let length = (x.powf(2.) + y.powf(2.)).sqrt();
+                let normalized_direction = Position {
+                    x: x / length,
+                    y: y / length,
+                };
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/map.rs]: this was deleted and changed to use the `Position::normalize` method

File: apps/arena/lib/arena/game_updater.ex (hunk 17 above)
[the remark this answers is not in the document]
Comment (Nico-Sanchez) [reply apps/arena/lib/arena/game_updater.ex:619]: This is because some obstacles doesn't have the vertices key, right? I think they all should, but if it's not needed it could be empty (nil).

File: apps/arena/native/physics/src/collision_detection/sat.rs (hunk 28 above)
[the remark this answers is not in the document]
Comment (Nico-Sanchez) [reply apps/arena/native/physics/src/collision_detection/sat.rs]: I meant the naming. It's much more clear now.

File: apps/arena/lib/arena/game_updater.ex (hunk 17 above)
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/lib/arena/game_updater.ex:619]: hmm right, something like an schema or an object, i can see that. though i deleted this, it was overloaded with parameters from the obstacle map so i sent the map and pattern match. Added "vertices" variable to circles too so all the obstacles have consistency

**Review by Nico-Sanchez — APPROVED** [review]

File: apps/arena/native/physics/src/collision_detection/sat.rs [hunk 30]
```diff
@@ -0,0 +1,279 @@
...
+        } else {
+            normal = axis;
+        }
+        result_depth = axis_depth;
+    }
+
+    (true, normal, result_depth)
+}
+
+// Uncomment this if we need a polygon-polygon collision detection
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/arena/native/physics/src/collision_detection/sat.rs:118]: This was something that i added as an extra since we can calculate this with this algorithm, we can fix it when we need polygon-polygon collisions (if we ever need them)

**Review by ncontinanza — APPROVED** [review]
